node/src/components/consensus/highway_core/state/endorsement_tracker.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::HashMap;
+
+use crate::components::consensus::{highway_core::validators::ValidatorIndex, traits::Context};
+
+use super::{Panorama, State};
+
+/// Helper structure used for computing which votes need endorsement.
+pub(crate) struct EndorsementTracker<'a, C: Context> {
+    /// Validators that are known to be equivocators by the initial panorama.
+    known_equivocators: Vec<ValidatorIndex>,
+    /// Visited units. Used for avoding visitng the same unit twice.
+    visited_nodes: Vec<C::Hash>,
+    /// Collection from a unit hash to indexes of equivocators that this unit cites as honest.
+    equiv_citations: HashMap<C::Hash, Vec<ValidatorIndex>>,
+    /// Read-only state.
+    state: &'a State<C>,
+}
+
+impl<'a, C: Context> EndorsementTracker<'a, C> {
+    pub(crate) fn new(state: &'a State<C>) -> Self {
+        EndorsementTracker {
+            known_equivocators: Vec::new(),
+            visited_nodes: Vec::new(),
+            equiv_citations: HashMap::new(),
+            state,
+        }
+    }
+
+    /// Returns votes that cite equivocating validators as honest.
+    pub(crate) fn run(mut self, panorama: &Panorama<C>) -> HashMap<C::Hash, Vec<ValidatorIndex>> {
+        self.known_equivocators = panorama.iter_faulty_validators().collect();
```

Comment:
As discussed, we probably need to distinguish between cited equivocations and other accusations here.
We might have to introduce `Observation::Equivocated`, and add a parameter to `Observation::Faulty(C::Hash)`, so the latter would mean that we have evidence against the validator but are not citing an equivocation and are seeing a unique latest unit?
For the purpose of the LNC, `Faulty(_)` and `Honest(_)` are probably equivalent.